### DISCUSSION TITLE
Implementation of inclination dependence of the GB energies through $\gamma$ in New phase-field model by Moelans in 2022

### DISCUSSION THREAD

**Discussion by yixishen** [idaholab/moose#26842] [Q&A Modules: Phase field]
Dear MOOSE community,

I am aiming to develop a phase-field (PF) model with a strong inclination-dependent anisotropic grain boundary (GB) energy and mobility. I understand that GBAnisotropy can be utilized for calculating PF parameters following the framework established by Moelans et al. in their 2008 Phys. Rev. B publication (Phys. Rev. B 78, 024113 – Published 16 July 2008). To my knowledge, the inclination dependence is incorporated via angle-dependent PF parameters such as $\kappa(\theta)$ and $\gamma(\theta)$, without necessitating additional terms in the evolution equations (specifically, the Allen-Cahn equation).

However, recent advancements by Moelans ([2022](https://doi.org/10.1016/j.matdes.2022.110592)) suggests that to have inclination dependent in a fully variational way, as $\kappa$($\theta$) is essentially a function of the gradient of order parameters $\eta_i$ and $\eta_j$, an extra term is introduced to the evolution equations:
$$- m \nabla \cdot \left( \sum_{j \neq i} \left( \frac{\partial \gamma_{ij}}{\partial n_i} \right) n_i^2 n_j ^2\right)$$

And my question is, does moose has any kernel that support this term? Or I may need to create my own function/application for this term?

Thanks
Yixi

**Comment by GiudGiud** [comment]
@laagesen @jiangwen84

**Reply by laagesen** [reply]
Hi Yixi- it looks like the extra term you are asking about is the last term in Equation 16 of the 2022 Moelans paper, right? Just wanted to clarify, because the derivative of gamma is actually with respect to grad eta in the paper, rather than with respect to eta as you wrote. We don't have any kernel developed for that type of dependence currently. If you wanted to implement it, you'd need to write a kernel. It would be similar to ACInterfaceKobayashi1 kernel used in the snow.i example: moose/modules/phase_field/examples/anisotropic_interfaces/snow.i
You might also need to write a kernel similar to ACInterfaceKobayashi2 to capture the correct Jacobians for the last term of the second line of Eq. 16. We would welcome a contribution if you would like to do this and are willing to help out along the way. I would suggest first figure out what functional form you want to use for \gamma_{i,j} and what the analytical forms are for the derivative (this might be in the supplemental material for the paper, I didn't double check it)

**Reply by yixishen** [reply]
Hi Larry,

Thanks for your response. Yes, the extra term is the last term in Eq 16 of 2022 Moelans paper. And thank you for lead me to ACInterfaceKobayashi1 kernel and  ACInterfaceKobayashi2 kernel. I would be happy to make some contribution to the community. Will work on it as your suggested and let you know if I have questions.

Thanks,
Yixi

**Reply by yixishen** [reply]
> Hi Yixi- it looks like the extra term you are asking about is the last term in Equation 16 of the 2022 Moelans paper, right? Just wanted to clarify, because the derivative of gamma is actually with respect to grad eta in the paper, rather than with respect to eta as you wrote. We don't have any kernel developed for that type of dependence currently. If you wanted to implement it, you'd need to write a kernel. It would be similar to ACInterfaceKobayashi1 kernel used in the snow.i example: moose/modules/phase_field/examples/anisotropic_interfaces/snow.i You might also need to write a kernel similar to ACInterfaceKobayashi2 to capture the correct Jacobians for the last term of the second line of Eq. 16. We would welcome a contribution if you would like to do this and are willing to help out along the way. I would suggest first figure out what functional form you want to use for \gamma_{i,j} and what the analytical forms are for the derivative (this might be in the supplemental material for the paper, I didn't double check it)

Hi Larry, 
Sorry for my delayed response, I just started working on this very recently. As I go through with ACInterfaceKobayashi1, and some other examples. I noticed there is also a source file "ADACInterfaceKobayashi1.C", which applies the Automatic Differentiation feature in MOOSE. If my understanding is right, AD implementation is encouraged and for AD, Jacobian is not required and only residuals is required. Is my understanding correct?
Thank you for your help,
Yixi

**Reply by laagesen** [reply]
Yes that is correct! We have not implemented AD versions of many of the phase-field kernels because we already have full Jacobians implemented. AD versions have been found to be slower than non-AD versions which have the full Jacobians implemented. However in cases where the Jacobians are so complex they haven't been implemented, the AD version performs better.

**Reply by yixishen** [reply]
Thanks for the confirmation. I’m testing the script I coded by creating a custom application as illustrated on the MOOSE website. To use other modules from MOOSE, I edited the Makefile and set ALL_MODULES := yes.

While testing the custom code, I found that the custom application works well on my local MacBook, though it takes a lot of time and almost gets stuck at “Setting up” when running on the UCSB cluster. For both the cluster and MacBook (apple chip), I installed MOOSE using the method described for [Linux and MacOS].

	1.	I think this issue is due to the custom code because if I don’t use that part on the cluster, it does not get stuck at “Setting up.”
	2.	I also tried to just copy and paste my header and src files to the ../modules/phase_field/ directory and recompiled the framework, but the issue persists.

I’m confused about why the exact same custom code works well on my local machine but not on the cluster. Could this be due to differences in the compiler or environment?

**Reply by GiudGiud** [reply]
What are the specs on each machine? Could it be that your local machine is just a lot faster

**Reply by yixishen** [reply]
> What are the specs on each machine? Could it be that your local machine is just a lot faster

My local machine is m3 pro chip, and the cluster is intel Xeon Gold 6248R processors (48 cores/node). For single core, I think m3 pro is faster than intel, but not a lot. I don't think that is the reason, because if I don't use my custom kernels/materials on cluster, it works well. If I use my custom kernels/materials, it will stuck at "setting up", which is like the initialization of the simulation?

I think it might due to some part of my code, I'm working on that.

Thanks,
Yixi

**Reply by GiudGiud** [reply]
It could be slower in parallel if the custom code is not efficient. Do you use all 48 cores on the cluster?

**Reply by yixishen** [reply]
> Yes that is correct! We have not implemented AD versions of many of the phase-field kernels because we already have full Jacobians implemented. AD versions have been found to be slower than non-AD versions which have the full Jacobians implemented. However in cases where the Jacobians are so complex they haven't been implemented, the AD version performs better.

I have some general questions. 
1), If I decided to use AD version, does that mean all other kernel should also be using an AD version? For example, the second term in the evolution equation,  $-Lm(\eta_i^3 - \eta_i + 2 \eta_i \sum_{j \neq i} \gamma_{ij} \eta_i^2)$, is embedded via kernel ACGrGrMulti. 
2), I noticed that the ACGrGrMulti.C source code has parts for both AD and non-AD version of ACGrGrMulti. By using "
registerMooseObject("PhaseFieldApp", ADACGrGrMulti); ", the ADACGrGrMulti is also register in MOOSE. Does that mean I can just use an AD version by defining the type to ADACGrGrMulti?

**Reply by GiudGiud** [reply]
1) no you can mix and match AD and non AD objects in the input file. The only thing is that some material properties might be required to be AD to build / use AD kernels

2) yes. When both are registered it means both versions are available, selectable through the type of the object in the input file

**Reply by yixishen** [reply]
Thank you for your quick response. I have grabbed some source code for the extra term and pushed to my own GitHub. I'm still testing and debugging. May I ask what is the properl if I would like you to take a look at these codes? 

Thanks
Yixi

**Reply by GiudGiud** [reply]
Hello

I don't know much about phase field. If you hit issues we will try to tag someone more knowledgeable.
The best way to share code is to make a github repository

**Comment by yixishen** [comment]
I test it with a very small model, so I only use 4 cores. In fact, the model is initiated at 200 by 200 mesh, one core is ok.On Jun 4, 2024, at 4:00 AM, Guillaume Giudicelli ***@***.***> wrote:﻿
It could be slower in parallel if the custom code is not efficient. Do you use all 48 cores on the cluster?

**Reply by GiudGiud** [reply]
Ok let's stick to one core then

Using a debugger (see debugging instructions on our website) on the cluster with one core, use the interrupt at regular intervals to find out in which line the code is spending a lot of time / getting stuck 

Alternatively, use a profiler (see profiling instructions) on the Mac to find the slow section
